typescript/sdk/src/token/EvmERC20WarpRouteReader.ts

```diff
@@ -261,87 +318,168 @@ export class EvmERC20WarpRouteReader extends HyperlaneReader {
-          warpRouteAddress,
-          this.provider,
-        );
-        xerc20Token = await hypXERC20Lockbox.xERC20();
-        token = xerc20Token;
-        lockbox = await hypXERC20Lockbox.lockbox();
-      } else {
-        const erc20 = HypERC20Collateral__factory.connect(
-          warpRouteAddress,
-          this.provider,
-        );
-        token = await erc20.wrappedToken();
-      }
+    const deriveFunction = this.deriveTokenConfigMap[type];
+    if (!deriveFunction) {
+      throw new Error(
+        `Provided unsupported token type "${type}" when fetching token metadata on chain "${this.chain}" at address "${warpRouteAddress}"`,
+      );
+    }
 
-      const { name, symbol, decimals } = await this.fetchERC20Metadata(token);
+    const config = await deriveFunction(warpRouteAddress);
+    return HypTokenConfigSchema.parse(config);
+  }
 
-      if (type === TokenType.XERC20 || type === TokenType.XERC20Lockbox) {
-        xERC20Metadata = await this.fetchXERC20Config(token, warpRouteAddress);
-      }
+  private async deriveHypXERC20TokenConfig(
+    hypTokenAddress: Address,
+  ): Promise<HypTokenConfig> {
+    const hypTokenInstance = HypXERC20__factory.connect(
+      hypTokenAddress,
+      this.provider,
+    );
 
-      return {
-        ...xERC20Metadata,
-        type,
-        name,
-        symbol,
-        decimals,
-        token: lockbox || token,
-      };
-    } else if (
-      type === TokenType.synthetic ||
-      type === TokenType.syntheticRebase
-    ) {
-      const baseMetadata = await this.fetchERC20Metadata(warpRouteAddress);
-
-      if (type === TokenType.syntheticRebase) {
-        const hypERC4626 = HypERC4626__factory.connect(
-          warpRouteAddress,
-          this.provider,
-        );
-        const collateralChainName = this.multiProvider.getChainName(
-          await hypERC4626.collateralDomain(),
-        );
-        return { type, ...baseMetadata, collateralChainName };
-      }
+    const collateralTokenAddress = await hypTokenInstance.wrappedToken();
+    const [erc20TokenMetadata, xERC20Metadata] = await Promise.all([
+      this.fetchERC20Metadata(collateralTokenAddress),
+      this.fetchXERC20Config(collateralTokenAddress, hypTokenAddress),
+    ]);
 
-      return { type, ...baseMetadata };
-    } else if (type === TokenType.native) {
-      const chainMetadata = this.multiProvider.getChainMetadata(this.chain);
-      if (chainMetadata.nativeToken) {
-        const { name, symbol, decimals } = chainMetadata.nativeToken;
-        return {
-          type,
-          name,
-          symbol,
-          decimals,
-        };
-      } else {
-        throw new Error(
-          `Warp route config specifies native token but chain metadata for ${this.chain} does not provide native token details`,
-        );
-      }
-    } else {
+    return {
+      ...erc20TokenMetadata,
+      type: TokenType.XERC20,
+      token: collateralTokenAddress,
+      xERC20: xERC20Metadata.xERC20,
```

Comment:
in the above Promise.all, if this.fetchXERC20Config() throws then it would return an empty object. that means the `xERC20` would be undefined. is this okay, or should fetchXERC20Config _only_ throw?

Reply:
only `fetchXERC20Config` should throw at the moment, we actually need to change something in there because we assume that all the XERC20 tokens comply with the same interface for the limits, which is not true, for example renzo's XERC20's don't have the same interface as velodrome's contracts

Reply:
discussed in dms too